File: ScalaQuickStart/src/main/java/com/tirthal/learning/java2scala/javaway/classobj/CashFlow.java
package com.tirthal.learning.java2scala.javaway.classobj;

// Java way - overloaded constructors 

public class CashFlow {

	// Fields
	private final Double amt;
	private final String curr;
	
	// Java's overloaded constructors
	public CashFlow(Double amt, String curr) {
		this.amt = amt;
		this.curr = curr;
	}
	
	public CashFlow(Double amt) {
		this(amt, "INR");
	}
	
	public CashFlow(String curr) {
		this(0.0, curr);
	}
	
	// Getter methods
	public Double getAmount() {
		return amt;
	}
	
	public String getCurrency() {
		return curr;
	}
	
	// Try it
	public static void main(String[] args) {
		CashFlow cf = new CashFlow(500.0);
		System.out.println("Amount=" + cf.getAmount());
		System.out.println("Currency=" + cf.getCurrency());
	}
}

File: ScalaQuickStart/src/main/scala/com/tirthal/learning/java2scala/scalaway/classobj/CashFlow.scala
package com.tirthal.learning.java2scala.scalaway.classobj

// Scala way - Primary and Auxiliary constructors

// class having a primary constructor that defines the parameters for the class
class CashFlow(amt: Double, curr: String) { // primary two-argument constructor

  // a class may have additional auxiliary constructors, which really act like providers of default parameters to the primary constructor
  def this(amt: Double) = this(amt, "INR")
  def this(curr: String) = this(0, curr)

  // define methods without including brackets or a parameter list (even an empty one)
  def amount = amt
  def currency = curr
}

// Try it
object CashFlow {
  def main(args: Array[String]) {
    
    val cf = new CashFlow(500.0)
    println("Amount=" + cf.amount)
    println("Currency=" + cf.currency)
    
  }
}